design/secretless-e2e-tests-dnd.md

```diff
@@ -0,0 +1,276 @@
+
+The `Service Engine` is a combination of steps defined in the `Test Runner`. The Engine orchestrates the lifecycle of a Service. It provides step definitions for starting Services, stopping services regardless of the test outcome, creating named pipes for the Service output and writing that to the test context.
+
+A `Test Case` is a sequence of steps that goes through a scenario to validate some given functianlity. It should really include at least one step making an assertion as part of the validation.
+
+A `Context` is a temporary store within a test case. Steps can read and write from the store. The key `service_name` will be used to identify this output.
+
+Here is an example of the components in this architecture working together.
+
+1. Service definitions exist for Conjur and MySQL
+    1. In this example it is MySQL and Conjur. As mentioned in the description of Service Definition, output from `start_command` is written into the test case `context` under a path corresponding to `service_name`. An example for MySQL is provided below.
+    ```
+    {
+      "service_name": "mysql_service",
+      "service_type": "myql",
+      "start_command": "mysql_start.sh", # imagine that "mysql_start.sh" calls out to "helm install"
+      "stop_command": "mysql_stop.sh"
+    }
+    ```
+1. Service Configs exist for Conjur and MySQL. These files would exist as `./mysql_config.json` and `./conjur_config.json`. An example for Conjur is provided below
+    ```
+    {
+      "conjur_account": "test",
+      "version": "10.9",
+      "docker_image": "registry.tld/conjur-appliance:10.9-stable"
+      "admin_user": "kumbi",
+      "admin_password": "definetelyasecret"
+    }
+    ```
+1. A service instances file exists for the test case under consideration. This file contains an array of Service Definition and Service Config pairs. This files would be called `services.json`
+    ```
+    [
+      {
+          "config": "./mysql_config.json",
+          "service_definition": "mysql_service",
+      },
+      {
+          "config": "./conjur_config.json",
+          "service_definition": "conjur_service"
+      }
+    ]
+    ```
+1. There is a test case written in Gherkin for validating using the MySQL service connector with Conjur. See below that there is an expliit step specifying the services that are expected to be running. 
+    ```
+    Feature: Secretless happy path with database connectors
+    
+      Scenario: Connection to MySQL with MySQL credentials stored in Conjur
+        Given services are running as defined in "./services.json" and the results are store in context path "services"
```

Comment:
This might need reduction in length. Maybe `Given running services "./services.json" and results in "some place"`?